geomstats/geometry/rank_k_psd_matrices.py

```diff
@@ -0,0 +1,209 @@
+r"""The manifold of Positive Semi Definite matrices of rank k PSD(n,k)."""
+
+import geomstats.backend as gs
+from geomstats.geometry.manifold import Manifold
+from geomstats.geometry.general_linear import GeneralLinear
+from geomstats.geometry.matrices import Matrices
+from geomstats.geometry.spd_matrices import (
+    SPDMatrices,
+    SPDMetricBuresWasserstein,
+    SPDMetricAffine,
+    SPDMetricEuclidean,
+    SPDMetricLogEuclidean,
+)
+from geomstats.geometry.symmetric_matrices import SymmetricMatrices
+
+
+class RankKPSDMatrices(Manifold):
+    r"""Class for PSD(n,k).
+
+    The manifold of symmetric positive definite (PSD) matrices of rank k.
+    Parameters
+    ----------
+    n : int
+        Integer representing the shape of the matrices: n x n.
+    k: int
+        Integer representing the rank of the matrix (k<n).
+    """
+    def __init__(
+        self,
+        n,
+        k,
+        **kwargs
+    ):
+        super(RankKPSDMatrices, self).__init__(**kwargs,
+                                               dim=int(k * n - k * (k + 1) / 2))
+        self.n = n
+        self.rank = k
+        self.sym = SymmetricMatrices(self.n)
+
+    def belongs(self, mat, atol=gs.atol):
+        r"""Check if the matrix belongs to the space.
+
+        Parameters
+        ----------
+        mat : array-like, shape=[..., n, n]
+            Matrix to be checked.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if mat is an SPD matrix.
+        """
+        is_symmetric = self.sym.belongs(mat, atol)
+        eigvalues = gs.linalg.eigvalsh(mat)
+        is_semipositive = gs.all(eigvalues > -atol, axis=-1)
+        is_rankk = gs.sum(gs.where(eigvalues < atol, 0, 1), axis=-1) == self.rank
+        belongs = gs.logical_and(
+            gs.logical_and(is_symmetric, is_semipositive), is_rankk
+        )
+        return belongs
+
+    def projection(self, point):
+        r"""Project a matrix to the space of PSD matrices of rank k.
+
+        First the symmetric part of point is computed,
+        then the matrix is multiplied by the I_\epsilon matrix.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n, n]
+            Matrix to project.
+
+        Returns
+        -------
+        projected: array-like, shape=[..., n, n]
+            PSD matrix rank k.
+        """
+
+        sym = Matrices.to_symmetric(point)
+        _, s, v = gs.linalg.svd(sym)
+        h = gs.matmul(Matrices.transpose(v), s[..., None] * v)
```

Comment:
I agree with you,  it seems like a redundant step to turn the matrix into a symmetric one (as we already do that in the line before). It is a step in the implementation in matlab take from this reference: https://www.sciencedirect.com/science/article/pii/0024379588902236
Let's discuss it in December because the function works but I am not sure if it is a smart solution and I haven't had much time to check for other smarter option.